### PR TITLE
Improve help for make command

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @steveklabnik

### CODE REVIEW

**Comment by rkruppe** [comment]
Some similar information is in [CONTRIBUTING.md in the "Build system" section](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#the-build-system). Perhaps these targets should be added there?

**Comment by GuillaumeGomez** [comment]
I think the makefile's readme should be on its own. However the point here is just to open the debate. :)

**Comment by alexcrichton** [comment]
Could this be merged in perhaps some other documentation? It'd be kinda unfortunate to have *another* top-level file dedicated to a system that's being phased out.

**Comment by GuillaumeGomez** [comment]
@alexcrichton: Absolutely, where would you want to add it? Any file in mind? The one @rkruppe proposed doesn't seem to be a very good fit.

**Comment by alexcrichton** [comment]
I believe we already have the header of `Makefile.in`, `make help`, and `make tips` lying around.

**Comment by steveklabnik** [comment]
r? @alexcrichton 

Since you care about where this is located, I'll let you take this one :smile:

**Comment by GuillaumeGomez** [comment]
Updated. So I extended `make help` and added a help message if you enter an invalid make rule.

File: Makefile.in [hunk 1]
```diff
@@ -257,3 +271,9 @@ ifneq ($(strip $(findstring TAGS.emacs,$(MAKECMDGOALS)) \
   CFG_INFO := $(info cfg: including ctags rules)
   include $(CFG_SRC_DIR)mk/ctags.mk
 endif
+
+.DEFAULT:
```
Comment (alexcrichton) [inline Makefile.in]: Does this break a bare `make`?

File: Makefile.in [hunk 2]
```diff
@@ -108,6 +108,20 @@
 #
 #     run `make nitty-gritty`
 #
+# # Make command examples
+#
+# ## Docs linked commands
+#
+# * make check-stage1-rustdocck: Builds rustdoc. It has the advantage to compile quite quickly since we're only using stage1 executables.
+# * make doc/error-index.md: Gets all doc blocks from doc comments and error explanations to put them in a markdown file. You can then test them by running "rustdoc --test error-index.md".
```
Comment (alexcrichton) [inline Makefile.in]: Can you be sure to wrap these to 80 characters?

File: Makefile.in [hunk 3]
```diff
@@ -108,6 +108,20 @@
...
+#
+# ## rustc linked commands
+#
+# * make rustc-stage2: Builds a functional rustc version.
+# * make check-stage2-cfail: Starts compile-fail tests.
```
Comment (alexcrichton) [inline Makefile.in]: This is already mentioned above in the "Tip Line" section

Comment (alexcrichton) [inline Makefile.in]: (both actually)

File: Makefile.in (hunk 1 above)
Comment (GuillaumeGomez) [inline Makefile.in]: It starts as expected.

**Comment by GuillaumeGomez** [comment]
Updated.

File: Makefile.in (hunk 1 above)
Comment (alexcrichton) [inline Makefile.in]: What does this end up doing? Is this just to translate `make foo` into a help message here?

Comment (GuillaumeGomez) [inline Makefile.in]: No, more than this. If you enter an invalid rules, this rule will be called.

**Pull request merged**